dan_layer/consensus/src/hotstuff/on_receive_proposal.rs

```diff
@@ -229,6 +229,57 @@ where TConsensusSpec: ConsensusSpec
         Ok(())
     }
 
+    fn on_receive_foreign_block(
+        &self,
+        tx: &mut <TConsensusSpec::StateStore as StateStore>::WriteTransaction<'_>,
+        block: &Block,
+        foreign_committee_shard: &CommitteeShard,
+    ) -> Result<(), HotStuffError> {
+        // Save the QCs if it doesnt exist already, we'll reference the QC in subsequent blocks
+        block.justify().save(tx)?;
+
+        // TODO(perf): n queries
+        for cmd in block.commands() {
+            let Some(t) = cmd.local_prepared() else {
+                continue;
+            };
+            let Some(mut tx_rec) = self.transaction_pool.get(tx, &t.id).optional()? else {
+                continue;
+            };
+
+            if tx_rec.stage().is_complete() {
+                warn!(
+                    target: LOG_TARGET,
+                    "⚠️ Foreign proposal received after transaction {} is COMPLETE. Ignoring.",
+                    tx_rec.transaction.id);
+                continue;
+            }
+
+            tx_rec.update_evidence(tx, foreign_committee_shard, *block.justify().id())?;
+            let change_to_abort = cmd.decision().is_abort() && tx_rec.original_decision().is_commit();
+            if change_to_abort {
+                info!(
+                    target: LOG_TARGET,
+                    "⚠️ Foreign shard ABORT {}. Update decision to ABORT",
+                    tx_rec.transaction.id
+                );
+                tx_rec.update_decision(tx, Decision::Abort)?;
```

Comment:
I don't think this is 100% correct. We should update the __pending__ decision to abort, but we only actually update the decision when the local leader proposes to us to abort.

In other words, we change the decision that we would make if we would become the leader and propose to change this transaction.

Interestingly, if we receive this foreign proposal and it looks like it's an Abort, but the leader proposes an Accept, we would have to check the evidence that they have provided that we should Accept, and if it checks out, it should override our local abort. This is only possible in the threat model of an entire committee being malicious so we should throw up a big warning flag and potentially even record this somewhere for reconciliation. It might even be worth stalling consensus at this point so that we don't mangle the data